### PR TITLE
[WIP] Initial work on running tests against the images for PRs

### PR DESCRIPTION
**Pull request by glennc**
This mostly copies the current state of the dotnet-docker repo, and probably should evolve in the same way.

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
+set -e 	# Exit immediately upon failure
+set -o pipefail  # Carry failures over pipes
+
+repo_root="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+docker_repo="microsoft/aspnetcore"
```
Comment (natemcmaster) [inline build.sh]: We need to also account for files that are on the `microsoft/aspnetcore-build` repo instead.

File: build.sh [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+function build_dockerfiles {
+    for dockerfile_dir in ${1}; do
+        echo "----- ${dockerfile_dir} -----"
+        tag="${docker_repo}:$( sed -e 's/.\///' -e 's/jessie\///' -e 's/\//-/g' <<< "${dockerfile_dir}" )"
+        echo "----- Building ${tag} -----"
+        docker build -t "${tag}" "${dockerfile_dir}"
+    done
+}
+
+build_dockerfiles "$( find . -path './.*' -prune -o -name 'Dockerfile' -a -path '**/*' -print0 | xargs -0 -n1 dirname )"
```
Comment (natemcmaster) [inline build.sh]: Nit: consider adding trailing end-line to all files.

File: test/create-run-publish-app.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+echo "Testing self-contained deployment"
+if [[ $2 == *"projectjson"* ]]; then
+    runtimes_section="  },\n  \"runtimes\": {\n    \"debian.8-x64\": {}\n  }"
+    sed -i '/"type": "platform"/d' ./project.json
+    sed -i "s/^  }$/${runtimes_section}/" ./project.json
+
+    dotnet restore
+    dotnet publish -o publish/self-contained
+else
+    sed -i '/<PropertyGroup>/a \    <RuntimeIdentifiers>debian.8-x64<\/RuntimeIdentifiers>' ./${PWD##*/}.csproj
```
Comment (natemcmaster) [inline test/create-run-publish-app.sh]: No need to edit the file. This can be based on command line as `dotnet restore /p:RuntimeIdentifiers=debian.8-x64`. Or, we can just add it to the file.

File: test/test.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env bash
+set -e 	# Exit immediately upon failure
+set -o pipefail  # Carry failures over pipes
+
+docker_repo="microsoft/aspnetcore"
```
Comment (natemcmaster) [inline test/test.sh]: `microsoft/aspnetcore-build`

File: build.sh (hunk 1 above)
Comment (glennc) [inline build.sh]: Because this doesn't actually push to dockerhub it doesn't matter. You just need some name to refer to the images that are built to test them.

File: test/create-run-publish-app.sh (hunk 3 above)
Comment (glennc) [inline test/create-run-publish-app.sh]: RIght, i didn't realize this was an option with the new CLI. So if a csproj doesn't specify the RID as long as I have restored specifying the RID I can also publish for a specific RID and it will be standalone?

Comment (natemcmaster) [inline test/create-run-publish-app.sh]: Yes.

File: build.sh (hunk 1 above)
Comment (natemcmaster) [inline build.sh]: In that case, could we use something different, like 'test/aspnetcore'?

File: test/create-run-publish-app.sh [hunk 5]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
```
Comment (MichaelSimons) [inline test/create-run-publish-app.sh]: @friism recently made a [contribution](https://github.com/dotnet/dotnet-docker/commit/6ce7bf25a74f8097e18757b484b4cd2c9399302c) to the dotnet-docker repo last week which changed this pattern to use a Dockerfile instead.  This removed the host dependency which allows the tests to be run against a remote daemon.

File: 1.1/jessie/runtime/Dockerfile [hunk 6]
```diff
@@ -6,8 +6,8 @@ ENV DOTNET_HOSTING_OPTIMIZATION_CACHE /packagescache
 COPY ./build-cache.sh /packagescache/build-cache.sh
 
 # set up package cache
-RUN chmod +x /packagescache/build-cache.sh \
-    && /packagescache/build-cache.sh https://dist.asp.net/packagecache/1.1.0/aspnetcore.packagecache-1.1.0-legacy-debian.8-x64.tar.gz \
+RUN chmod +x /packagescache/build-cache.sh
+RUN /packagescache/build-cache.sh https://dist.asp.net/packagecache/1.1.0/aspnetcore.packagecache-1.1.0-legacy-debian.8-x64.tar.gz \
```
Comment (friism) [inline 1.1/jessie/runtime/Dockerfile]: why this change? (did you intend to check this in?)

Comment (natemcmaster) [inline 1.1/jessie/runtime/Dockerfile]: IIRC this was changed because Glenn hit https://github.com/docker/docker/issues/9547

File: test/create-run-publish-app.sh (hunk 5 above)
Comment (glennc) [inline test/create-run-publish-app.sh]: Yes, my I have looked at his PR and intent to do the same here. But I already had enough of this done that it seemed worthwhile to do a PR and then change it instead of waiting. More coverage sooner seemed like a good idea.

File: 1.1/jessie/runtime/Dockerfile (hunk 6 above)
Comment (glennc) [inline 1.1/jessie/runtime/Dockerfile]: Yeah, the issue Nate linked is the reason. Presumably we don't have to chmod inside the container at all, and I had intended to look at removing it completely if it was possible.

Comment (friism) [inline 1.1/jessie/runtime/Dockerfile]: Ah, sorry

**Comment by natemcmaster** [comment]
Btw, travis is enabled now. Try pushing and empty commit to trigger a run.

**Comment by natemcmaster** [comment]
Tentative :shipit:. :watch: to merge until this actually passes on travis CI.

**Comment by natemcmaster** [comment]
CI Passed 🎆 Squash and :shipit:
